### PR TITLE
[Sampler] Metapath sampler for metapath2vec

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Provides the (cyclic) metapath-based random walk sampler required by metapath2vec.

This implementation takes 857 seconds to generate a corpus of the same size as that in Metapath2vec (which takes 1563 seconds) on AMiner.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

### CODE REVIEW

File: python/dgl/contrib/sampling/randomwalk.py [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
+from ... import ndarray
 
 __all__ = ['random_walk',
            'random_walk_with_restart',
            'bipartite_single_sided_random_walk_with_restart',
+           'metapath_cycle_random_walk',
            ]
 
+@register_object('sampler.RandomWalkTraces')
+class RandomWalkTraces(ObjectBase):
```
Comment (jermainewang) [inline python/dgl/contrib/sampling/randomwalk.py]: Briefly document the available attributes and types.

File: python/dgl/contrib/sampling/randomwalk.py [hunk 2]
```diff
@@ -165,4 +169,52 @@ def bipartite_single_sided_random_walk_with_restart(
             int(max_visit_counts), int(max_frequent_visited_nodes))
     return _split_traces(traces)
 
-_init_api('dgl.randomwalk', __name__)
+
+def metapath_cycle_random_walk(hg, etypes, seeds, num_traces, max_cycles):
```
Comment (jermainewang) [inline python/dgl/contrib/sampling/randomwalk.py]: Could we let user directly give a full metapath sequence so there is no "cycle" behavior in this API? User could directly give `[A,B,A,B,A,B]` and we would follow the metapath but could stop earlier. Also, change `etypes` argument to `metapath`.

File: python/dgl/contrib/sampling/randomwalk.py [hunk 3]
```diff
@@ -165,4 +169,52 @@ def bipartite_single_sided_random_walk_with_restart(
...
+    Returns
+    -------
+    traces : list[list[Tensor]]
+        traces[i][j] is the j-th trace generated for i-th seed.
+        traces[i][j][k] would have node type the same as the destination node type of edge
+        type ``etypes[k % len(etypes)]``
+
+    Notes
+    -----
+    The traces does **not** include the seed nodes themselves.
```
Comment (jermainewang) [inline python/dgl/contrib/sampling/randomwalk.py]: Is this a convention?

File: python/dgl/convert.py [hunk 4]
```diff
@@ -264,7 +264,7 @@ def hetero_from_relations(rel_graphs):
         retg._edge_frames[i].update(rgrh._edge_frames[0])
     return retg
 
-def to_hetero(G, ntypes, etypes, ntype_field=NTYPE, etype_field=ETYPE, metagraph=None):
+def to_hetero(G, ntypes, etypes, ntype_field=NTYPE, etype_field=ETYPE):
```
Comment (jermainewang) [inline python/dgl/convert.py]: Why revert these back?

File: src/graph/sampler/metapath.cc [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  const auto metagraph = hg->meta_graph();
+  uint64_t num_etypes = etypes->shape[0];
+  uint64_t num_seeds = seeds->shape[0];
+  const dgl_type_t *etype_data = static_cast<dgl_type_t *>(etypes->data);
+  const dgl_id_t *seed_data = static_cast<dgl_id_t *>(seeds->data);
+
+  std::vector<dgl_id_t> vertices;
+  std::vector<size_t> trace_lengths, trace_counts;
+
+  for (uint64_t seed_id = 0; seed_id < num_seeds; ++seed_id) {
```
Comment (jermainewang) [inline src/graph/sampler/metapath.cc]: Consider using omp to embarrassingly parallelize this loop? If so, you might need to trigger `SuccVec` first to materialize the lazy component before entering the critical area. You might also need to split the `vertices` vector into ones for each seed.

File: python/dgl/contrib/sampling/randomwalk.py (hunk 3 above)
Comment (BarclayII) [reply python/dgl/contrib/sampling/randomwalk.py]: Yes

File: src/graph/sampler/metapath.cc (hunk 5 above)
Comment (BarclayII) [reply src/graph/sampler/metapath.cc]: Sounds not very trivial to implement...  Can I postpone OpenMP parallelization to 0.5?

File: python/dgl/contrib/sampling/randomwalk.py (hunk 2 above)
Comment (BarclayII) [reply python/dgl/contrib/sampling/randomwalk.py]: Good point.

File: src/graph/sampler/metapath.cc (hunk 5 above)
Comment (jermainewang) [reply src/graph/sampler/metapath.cc]: sure. added a TODO note there

**Pull request merged**